compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -692,8 +694,27 @@ class Semantic {
   }
 
   /** Evaluate a list of expressions */
-  def eval(exprs: List[Tree], thisV: Addr, klass: ClassSymbol): Contextual[List[Result]] =
-    exprs.map { expr => eval(expr, thisV, klass) }
+  def eval(exprs: List[Tree], thisV: Addr, klass: ClassSymbol): Contextual[(List[Result], Env)] = exprs match {
+    case Nil => (Nil, env)
+    case h :: t => h match {
+      case v: ValDef => {
+        val res = eval(h, thisV, klass)
+        val newEnv = 
+          if res.value.promote("Try early promotion", h).isEmpty then Env(Map(v.symbol -> Hot)) else Env(Map(v.symbol -> res.value))
+        withEnv(env.union(newEnv)) {
+          val (res2, env2) = eval(t, thisV, klass)
+          (res :: res2, env2)
+        }
+      }
+      case _ => {
+        val res = eval(h, thisV, klass)
+        withEnv(env) {
+          val (res2, env2) = eval(t, thisV, klass)
+          (res :: res2, env2)
+        }
+      }
+    }
+  }
```

Comment:
Thanks @EnzeXing . If we do it in the place where we give semantics to local variables, then we can reduce the total changes to just 2 or 3 lines.

Reply:
This would be `case ValDef` in `def cases`.

Reply:
Sorry, I was not clear, it's in the use site of variables, i.e. `case TermRef(NoPrefix, _) =>`, not the definition site.

Reply:
You mean move the updating env part to the use site of variables?

Reply:
We don't need to update `env`. The idea is to evaluate its initializer when the variable is accessed -- and it's automatically cached.

Reply:
@liufengyun Sorry but I'm still a bit confused. My idea is evaluate the local variables when they are initialized and stored it, so that we can just look up when accessing it. How can we evaluate its initializer when it's accessed?

Reply:
> @liufengyun Sorry but I'm still a bit confused. My idea is evaluate the local variables when they are initialized and stored it, so that we can just look up when accessing it. How can we evaluate its initializer when it's accessed?

When the variable is accessed, rather than look up its value in env, we evaluate the rhs of the definition (ValDef) of the variable right there, when it's accessed.

Reply:
Suppose we have val v = new A in some local method, and the method accesses and returns v in the end. When we access v in the end, all we know about v is that it's a TermRef, but how can we know its rhs definition at this point? I thought we analyze val v = new A first and then accessing v at last, but how do we go the other way?

Reply:
From the `TermRef`, we should be able to get its `Symbol`, and from the `Symbol`, call `defTree` to find the `ValDef` where that variable was defined.

We need to make sure that we're doing this only for `TermRef`s that represent local variables. In particular, they should have `NoPrefix`, but there are probably additional conditions to check to ensure that it's a local variable.

Reply:
> From the `TermRef`, we should be able to get its `Symbol`, and from the `Symbol`, call `defTree` to find the `ValDef` where that variable was defined.
> 
> We need to make sure that we're doing this only for `TermRef`s that represent local variables. In particular, they should have `NoPrefix`, but there are probably additional conditions to check to ensure that it's a local variable.

Yes, that's exactly what I have in mind. Note that in `case ValDef`, the rhs is already evaluated and thus cached (need to add `cache = true`). The `eval` operation is actually a retrieval from the cache.

We need to set up the correct `thisV` for evaluating `rhs`, here is an example for local methods:

https://github.com/lampepfl/dotty/blob/4aa7f909f006a012cb78662b938ceda2059e67be/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala#L757-L760

Two corner cases to think about:

- invent an example where `thisV = Cold`. Just report a warning in this case.
- invent an example where `env != Nil`. 

There is a bug in handling local methods directly inside constructors called from local classes inside constructors: its `env` is not set up correctly. We can discuss a fix in next meeting.